include/cuco/detail/probe_sequences.cuh

```diff
@@ -0,0 +1,150 @@
+          typename Hash1           = cuco::detail::MurmurHash3_32<Key>,
+          typename Hash2           = cuco::detail::MurmurHash3_32<Key>,
+          cuda::thread_scope Scope = cuda::thread_scope_device>
+class double_hashing : public probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope> {
+ public:
+  using iterator = typename probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::iterator;
+  using probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::capacity_;
+  using probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::slots_;
+  using probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::cg_size;
+  using probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::is_vector_load;
+
+  __host__ __device__ explicit double_hashing(iterator slots, std::size_t capacity) noexcept
+    : probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>{slots, capacity}
+  {
+  }
+
+  template <typename CG>
+  __device__ iterator initial_slot(CG const& g, Key const k) noexcept
+  {
+    std::size_t index;
+    if constexpr (IsVectorLoad) {
+      step_size_ = (hash2_(k + 1) % (capacity_ / (cg_size() * 2) - 1) + 1) * cg_size() * 2;
+      index      = hash1_(k) % (capacity_ / (cg_size() * 2)) * cg_size() * 2 + g.thread_rank() * 2;
+    } else {
+      step_size_ = (hash2_(k + 1) % (capacity_ / cg_size() - 1) + 1) * cg_size();
+      index      = (hash1_(k) + g.thread_rank()) % capacity_;
+    }
+    return slots_ + index;
+  }
+
+  __device__ iterator next_slot(iterator s) noexcept
+  {
+    std::size_t index = s - slots_;
+    return &slots_[(index + step_size_) % capacity_];
+  }
+
+ private:
+  std::size_t step_size_;
+  Hash1 hash1_{};
+  Hash2 hash2_{};
+};  // class double_hashing
+
+template <typename Key,
+          typename Value,
+          uint32_t CGSize          = 8,
+          bool IsVectorLoad        = sizeof(Key) == 4,
+          typename Hash            = cuco::detail::MurmurHash3_32<Key>,
+          cuda::thread_scope Scope = cuda::thread_scope_device>
+class linear_probing : public probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope> {
+ public:
+  using iterator = typename probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::iterator;
+  using probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::capacity_;
+  using probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::slots_;
+  using probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::cg_size;
+  using probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>::is_vector_load;
+
+  __host__ __device__ explicit linear_probing(iterator slots, std::size_t capacity)
+    : probe_sequence_base<Key, Value, CGSize, IsVectorLoad, Scope>{slots, capacity}
+  {
+  }
+
+  template <typename CG>
+  __device__ iterator initial_slot(CG const& g, Key const k) noexcept
+  {
+    if constexpr (IsVectorLoad) {
+      return &slots_[(hash_(k) + g.thread_rank() * 2) % capacity_];
```

Comment:
Good point. The current vector load was initially proposed for 4B/4B use cases only since 16B load is the widest load instruction supported at the SASS level. As a result, each thread loads two 4B/4B consecutive slots (thus 16B in total). I think we are interested in this particular combination because it's the most common use case in cudf join (@jrhemstad correct me if I'm mistaken). Wider vector load support for other combinations (2B/4B, 2B/2B, etc.) can be definitely added in the future. For the current PR, I will add more documentation and asserts to make sure it's working as expected.

Reply:
OK, makes sense, thanks.